File: tasklens-core-kmp/src/commonMain/kotlin/dev/shushant/tasklens/diagnosis/DiagnosisRule.kt
package dev.shushant.tasklens.diagnosis

import dev.shushant.tasklens.core.DiagnosisClassification
import dev.shushant.tasklens.core.DiagnosisConfidence
import dev.shushant.tasklens.core.EnvironmentSnapshot
import dev.shushant.tasklens.core.Evidence
import dev.shushant.tasklens.core.ExecutionAttempt
import dev.shushant.tasklens.core.PlatformLimitation
import dev.shushant.tasklens.core.PossibleFactor
import dev.shushant.tasklens.core.ScheduledWork
import dev.shushant.tasklens.core.TaskLensEvent

data class DiagnosisContext(
    val task: ScheduledWork,
    val attempts: List<ExecutionAttempt>,
    val events: List<TaskLensEvent>,
    val environment: List<EnvironmentSnapshot> = emptyList()
)

data class DiagnosisCandidate(
    val ruleId: String,
    val title: String,
    val summary: String,
    val classification: DiagnosisClassification,
    val confidence: DiagnosisConfidence,
    val evidence: List<Evidence>,
    val possibleFactors: List<PossibleFactor> = emptyList(),
    val limitations: List<PlatformLimitation> = emptyList(),
    val priority: Int = 0
)

interface DiagnosisRule {
    val id: String
    val name: String
    /** Lower number = higher priority; rules with lower priority run first. Default 100. */
    val priority: Int get() = 100
    fun evaluate(context: DiagnosisContext): DiagnosisCandidate?
}

File: Sources/TaskLensDiagnosis/DiagnosisRule.swift
import Foundation
import TaskLensCore

public struct DiagnosisContext {
    public let task: ScheduledWork
    public let attempts: [ExecutionAttempt]
    public let events: [TaskLensEvent]

    public init(
        task: ScheduledWork,
        attempts: [ExecutionAttempt] = [],
        events: [TaskLensEvent] = []
    ) {
        self.task = task
        self.attempts = attempts
        self.events = events
    }
}

public struct DiagnosisCandidate {
    public let ruleId: String
    public let title: String
    public let summary: String
    public let classification: DiagnosisClassification
    public let confidence: DiagnosisConfidence
    public let evidence: [Evidence]
    public let possibleFactors: [PossibleFactor]
    public let limitations: [PlatformLimitation]
    public let priority: Int

    public init(
        ruleId: String,
        title: String,
        summary: String,
        classification: DiagnosisClassification,
        confidence: DiagnosisConfidence,
        evidence: [Evidence] = [],
        possibleFactors: [PossibleFactor] = [],
        limitations: [PlatformLimitation] = [],
        priority: Int = 0
    ) {
        self.ruleId = ruleId
        self.title = title
        self.summary = summary
        self.classification = classification
        self.confidence = confidence
        self.evidence = evidence
        self.possibleFactors = possibleFactors
        self.limitations = limitations
        self.priority = priority
    }
}

public protocol DiagnosisRule {
    var id: String { get }
    var name: String { get }
    func evaluate(context: DiagnosisContext) -> DiagnosisCandidate?
}
